Add SavePolarizedNXCanSAS to SANSSave

cailafinn opened the pull request:
### Description of work

#### Summary of work
Allows SANSSave (the SANS workflow algorithm for saving out reduced workspaces in a variety of formats) to make use of the new SavePolarizedNXCanSAS algorithm when the reduced output workspaces are in the polarized form. 

Part of the Polarized SANS Epic.

Fixes #38523 


#### Further detail of work
The algorithm now has an additional "PolarizationProps" property, which allows the passing of a dictionary containing all of the mandatory properties required for SavePolarizedNXCanSAS. 

The algorithm has also been adjusted to allow for the explicit passing of Workspace Groups. Groups are now processed as-is rather than attempting to process them one-by-one. Additional validation has been added to ensure that only algorithms that support such workspaces are allowed to be selected when calling the algorithm, and that workspaces explicitly are groups when calling for SavePolarizedNXCanSAS. 

### To test:

Use this script as a start, then have a play and try to break it.
```python
workspace_0 = CreateSimulationWorkspace(Instrument="LARMOR", BinParams="1,10,1000", UnitX="MomentumTransfer")
workspace_1 = CreateSimulationWorkspace(Instrument="LARMOR", BinParams="1,10,1000", UnitX="MomentumTransfer")
workspace_list = [workspace_0, workspace_1]
workspace = GroupWorkspaces(workspace_list)
workspace = ConvertSpectrumAxis(InputWorkspace=workspace, Target="ElasticQ", EFixed=1)

file_name = os.path.join(mantid.config.getString("defaultsave.directory"), "sample_sans_save_file")
pol_props = {
    "InputSpinStates": "+10,-10",
    "PolarizerComponentName": "a",
    "AnalyzerComponentName": "b",
    "FlipperComponentNames": "c, d, e",
    "MagneticFieldStrengthLogName": "f",
    "MagneticFieldDirection": "1,2,3",
}

# Act
SANSSave(workspace, file_name, NXcanSAS=True, PolarizedNXCanSAS=True, PolarizationProps=pol_props)
```

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

cailafinn committed:
Add polarization property

RE #38523

cailafinn committed:
Check mandatory inputs

RE #38523

cailafinn committed:
Implement polarized property

RE #38523

cailafinn committed:
Modernise save strategy selection

We can use match/case statements in python 3.10!

RE #38523

cailafinn committed:
Add tests and refactor validation

RE #38523

cailafinn committed:
Implement call to PolarizedNXCanSAS

RE #38523

cailafinn committed:
Allow saving of group workspaces with other save methods

RE #38523

cailafinn committed:
Add checking for group ws for incompatible algorithms

RE #38523

cailafinn committed:
Remove superfluous TODO comment

RE #38523

cailafinn committed:
Fix additional error return

RE #38523

cailafinn committed:
Add a release note

RE #38523

adriazalvarez commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSSave.py:
```diff
@@ -50,54 +69,49 @@ def PyInit(self):
 
         self.setPropertyGroup("Nexus", "FileFormats")
         self.setPropertyGroup("CanSAS", "FileFormats")
         self.setPropertyGroup("NXCanSAS", "FileFormats")
```
Not affecting anything, but could we rename this:
```suggestion
        self.setPropertyGroup("NXcanSAS", "FileFormats")
```

adriazalvarez commented on Testing/SystemTests/tests/framework/ISIS/SANS/WORKFLOWS/SANSSaveTest.py:
```diff
@@ -211,6 +212,123 @@ def test_that_zero_error_is_removed(self):
+        SANSSave(workspace, file_name, Nexus=True, NXcanSAS=True)
+
+    def test_group_workspaces_throw_for_incompatible_alg(self):
+        # Arrange
+        workspace_0 = CreateSimulationWorkspace(Instrument="LARMOR", BinParams="1,10,1000", UnitX="MomentumTransfer")
+        workspace_1 = CreateSimulationWorkspace(Instrument="LARMOR", BinParams="1,10,1000", UnitX="MomentumTransfer")
+        workspace_list = [workspace_0, workspace_1]
+        workspace = GroupWorkspaces(workspace_list)
+        workspace = ConvertSpectrumAxis(InputWorkspace=workspace, Target="ElasticQ", EFixed=1)
```
These lines to create a group workspace compatible with polarized save algorithm are repeated in few test, could we add a small helper function to create the group instead?

adriazalvarez commented on Testing/SystemTests/tests/framework/ISIS/SANS/WORKFLOWS/SANSSaveTest.py:
```diff
@@ -211,6 +212,123 @@ def test_that_zero_error_is_removed(self):
+                f"These properties are missing: {{'{list(pol_props.keys())[i]}'}}",
+            ):
+                # Act
+                SANSSave(
```
I guess this expects that the algorithm does not throw, should we add a small check that the file has been saved, and delete it after?

adriazalvarez reviewed (changes requested):
I have tried the algorithm with the test code and everything is working well and as expected. The code, in my impression, is well organized and clear and the inclusion of the new algorithm is well implemented. 
I have added some small comments/suggestions, particularly in tests. 

The only issue I have with the implementation is that I don't think we should enforce some properties as mandatory. The only mandatory properties should be that the input workspace is a group and the input spin states. The algorithm can save the data without the rest of the properties and in some experiments there may not be a polarizer or analyser (depending on the experiment). Also the magnetic field direction in my impression must be supplied by scientists, I don't know if they will always have the data. 
I would still add these other properties as additional properties, but not mandatory, and change the code accordingly (and the tests)

cailafinn commented:
> I don't think we should enforce some properties as mandatory. The only mandatory properties should be that the input workspace is a group and the input spin states.

Ah, I think those are left over from an older implementation of SavePolarizedNXCanSAS where the properties were split between mandatory and optional ones. I'll get that changed.

cailafinn committed:
Remove redundant mandatory properties

RE #38523

cailafinn committed:
Make naming of NXcanSAS consistent.

The correct format of NXcanSAS has a lowercase can. Make this the place
throughout the project to avoid confusion.

RE #38523

cailafinn committed:
Refactor to simplify group workspace creation

RE #38523

cailafinn committed:
Add check algorithm actually saved the file

RE #38523

adriazalvarez reviewed (changes requested):
Thanks for the changes! 
On my side, everything is working as it should. 
Just noticed a very small typo, but if you prefer to fix in a future pr to about rerunning the CI I can just approve it.

cailafinn commented on Testing/SystemTests/tests/framework/ISIS/SANS/WORKFLOWS/SANSSaveTest.py:
```diff
@@ -211,6 +222,102 @@ def test_that_zero_error_is_removed(self):
+
+        file_name = os.path.join(mantid.config.getString("defaultsave.directory"), "sample_sans_save_file")
+        pol_props = {
+            "InputSpinStatez": "0,1",
```
This is on purpose. I can add a comment to that effect, but it's intended to catch the case where `InputSpinStates` isn't defined (because of a typo, for instance), but there still has to be something in the dict object for it to pass the truthiness check.

MialLewis replied:
My aswell change this to `placeholder_prop` if making other changes.

MialLewis commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSSave.py:
```diff
@@ -204,12 +233,56 @@ def PyExec(self):
+            self._validate_workspace(errors, input_workspace.getItem(i))
+        if self.getProperty("PolarizedNXcanSAS").value:
+            self._validate_pol_props(errors)
+        if (
```
Does it make sense to check this and the props before validating all of the workspaces?

MialLewis commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSSave.py:
```diff
@@ -204,12 +233,56 @@ def PyExec(self):
+            additional_properties["Geometry"] = maybe_geometry.value
+
+        additional_properties["BackgroundSubtractionWorkspace"] = self.getProperty("BackgroundSubtractionWorkspace").value
+        additional_properties["BackgroundSubtractionScaleFactor"] = self.getProperty("BackgroundSubtractionScaleFactor").value
```
Can you either rename this function and bring:
```
        additional_properties["Transmission"] = transmission
        additional_properties["TransmissionCan"] = transmission_can
            if transmission:
                transmission = get_zero_error_free_workspace(transmission)
            if transmission_can:
                transmission_can = get_zero_error_free_workspace(transmission_can)
```
inside, or move these out of the function.

cailafinn committed:
Refactor collation of additional properties

RE #38523

cailafinn committed:
Rename confusing property

RE #38523

MialLewis reviewed (approved):
Thanks for the changes, code looking good.
